starfish/imagestack/parser/tileset/_parser.py

```diff
@@ -2,14 +2,101 @@
 This module parses and retains the extras metadata attached to TileSet extras.
 """
 import warnings
-from typing import Collection, MutableMapping, Tuple
+from typing import Collection, Mapping, MutableMapping, Optional, Tuple
 
+import numpy as np
 from slicedimage import Tile, TileSet
 
 from starfish.errors import DataFormatWarning
 from starfish.imagestack.dataorder import AXES_DATA
-from starfish.imagestack.parser import TileCollectionData, TileKey
-from starfish.types import Indices
+from starfish.imagestack.parser import TileCollectionData, TileData, TileKey
+from starfish.types import Coordinates, Indices, Number
+
+
+class _Expectations:
+    """
+    This class tracks all the tile shapes and the dtypes seen thus far during the decode of a
+    :py:class:`slicedimage.TileSet`.  If the shapes are not all identical, it will trigger a
+    ValueError.  If the kind of dtypes are not all identical, it will trigger a TypeError.
+    Additionally, if the dtypes are all of the same kind, but not all of the same size, it will
+    trigger a :py:class:`starfish.errors.DataFormatWarning`.
+    """
+    def __init__(self) -> None:
+        self.tile_shape: Optional[Tuple[int, ...]] = None
+        self.kind = None
+        self.dtype_size = None
+
+    def report_tile_shape(self, r: int, ch: int, z: int, tile_shape: Tuple[int, ...]) -> None:
+        if self.tile_shape is not None and self.tile_shape != tile_shape:
+            raise ValueError("Starfish does not support tiles that are not identical in shape")
+        self.tile_shape = tile_shape
+
+    def report_dtype(self, r: int, ch: int, z: int, dtype) -> None:
+        if self.kind is not None and self.kind != dtype.kind:
+            raise TypeError("All tiles should have the same kind of dtype")
+        if self.dtype_size is not None and self.dtype_size != dtype.itemsize:
+            warnings.warn(
+                f"Tile (R: {r} C: {ch} Z: {z}) has dtype {dtype}, which is different from one or "
+                f"more of the toher tiles.",
+                DataFormatWarning)
+
+        self.kind = dtype.kind
+        self.dtype_size = dtype.itemsize
+
+
+class SlicedImageTile(TileData):
+    """
+    This wraps a :py:class:`slicedimage.Tile`.  The key behavioral difference between this and
```

Comment:
This is the key difference -- what are the other differences? Might be good to spell them out. I might also transfer some of the logic from the PR description here -- we need this to efficiently load data, etc etc.